### PR TITLE
consolidate MaterializeResult tests

### PR DESCRIPTION
**Pull request by jamiedemaria**
## Summary & Motivation
Moves all of the MaterializeResult tests to one file. Adds new (skipped) tests to demonstrate direct invocation failure so that we can solve that code path in future work

### CODE REVIEW

**Comment by jamiedemaria** [comment]
Current dependencies on/for this PR:
* master
  * **PR #16920** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/16920" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/16920?utm_source=stack-comment).

**Review by schrockn — APPROVED** [review]
[yaytests](https://app.graphite.dev/macros/dagster-io)

File: python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+            asset_key="two", metadata={"baz": "qux"}
+        )
+
+    res = outs_multi_asset()
+    assert res[0].metadata["foo"] == "bar"
+    assert res[1].metadata["baz"] == "qux"
+
+
+@pytest.mark.skip(
+    "Direct invocation causes errors, unskip these tests once direct invocation path is fixed."
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py:314]: Link to issue # for extra credit

File: python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+        yield MaterializeResult(metadata={"foo": "bar"})
+
+    materialize([generator_asset], resources={"io_manager": TestingIOManager()})
+
+
+pytest.mark.skip(
+    "Direct invocation causes errors, unskip these tests once direct invocation path is fixed."
+)
+
+
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py:283]: ???

File: python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    def multi_asset_with_specs_and_no_type_annotation():
+        return MaterializeResult(asset_key="one"), MaterializeResult(asset_key="two")
+
+    assert materialize(
+        [multi_asset_with_specs_and_no_type_annotation],
+        resources={"io_manager": TestingIOManager()},
+    ).success
+
+
+@pytest.mark.skip("generator return types are interpreted as Any.")
```
Comment (schrockn) [inline python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py:262]: issue # for tracking

File: python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py (hunk 2 above)
Comment (jamiedemaria) [reply python_modules/dagster/dagster_tests/definitions_tests/test_materialize_result.py]: missing @ made black freak out

**Pull request merged**